src/vm-vcpu-ref/src/x86_64/msrs.rs

```diff
@@ -60,9 +93,177 @@ pub fn create_boot_msr_entries() -> Result<Msrs> {
+    MSR_RANGE!(
+        // IA32_MTRR_PHYSBASE0
+        0x200, 0x100
+    ),
+    MSR_RANGE!(
+        // MSR_CORE_C3_RESIDENCY
+        // MSR_CORE_C6_RESIDENCY
+        // MSR_CORE_C7_RESIDENCY
+        MSR_CORE_C3_RESIDENCY,
+        3
+    ),
+    MSR_RANGE!(MSR_IA32_MC0_CTL, 0x80),
+    SINGLE_MSR!(MSR_RAPL_POWER_UNIT),
+    MSR_RANGE!(
+        // MSR_PKGC3_IRTL
+        // MSR_PKGC6_IRTL
+        // MSR_PKGC7_IRTL
+        MSR_PKGC3_IRTL,
+        3
+    ),
+    SINGLE_MSR!(MSR_PKG_POWER_LIMIT),
+    SINGLE_MSR!(MSR_PKG_ENERGY_STATUS),
+    SINGLE_MSR!(MSR_PKG_PERF_STATUS),
+    SINGLE_MSR!(MSR_PKG_POWER_INFO),
+    SINGLE_MSR!(MSR_DRAM_POWER_LIMIT),
+    SINGLE_MSR!(MSR_DRAM_ENERGY_STATUS),
+    SINGLE_MSR!(MSR_DRAM_PERF_STATUS),
+    SINGLE_MSR!(MSR_DRAM_POWER_INFO),
+    SINGLE_MSR!(MSR_CONFIG_TDP_NOMINAL),
+    SINGLE_MSR!(MSR_CONFIG_TDP_LEVEL_1),
+    SINGLE_MSR!(MSR_CONFIG_TDP_LEVEL_2),
+    SINGLE_MSR!(MSR_CONFIG_TDP_CONTROL),
+    SINGLE_MSR!(MSR_TURBO_ACTIVATION_RATIO),
+    SINGLE_MSR!(MSR_IA32_TSCDEADLINE),
+    MSR_RANGE!(APIC_BASE_MSR, APIC_MSR_INDEXES),
+    SINGLE_MSR!(MSR_IA32_BNDCFGS),
+    SINGLE_MSR!(MSR_KVM_WALL_CLOCK_NEW),
+    SINGLE_MSR!(MSR_KVM_SYSTEM_TIME_NEW),
+    SINGLE_MSR!(MSR_KVM_ASYNC_PF_EN),
+    SINGLE_MSR!(MSR_KVM_STEAL_TIME),
+    SINGLE_MSR!(MSR_KVM_PV_EOI_EN),
+    SINGLE_MSR!(MSR_EFER),
+    SINGLE_MSR!(MSR_STAR),
+    SINGLE_MSR!(MSR_LSTAR),
+    SINGLE_MSR!(MSR_CSTAR),
+    SINGLE_MSR!(MSR_SYSCALL_MASK),
+    SINGLE_MSR!(MSR_FS_BASE),
+    SINGLE_MSR!(MSR_GS_BASE),
+    SINGLE_MSR!(MSR_KERNEL_GS_BASE),
+    SINGLE_MSR!(MSR_TSC_AUX),
+];
+
+/// Specifies whether a particular MSR should be included in vcpu serialization.
+///
+/// # Arguments
+///
+/// * `index` - The index of the MSR that is checked whether it's needed for serialization.
+fn msr_should_serialize(index: u32) -> bool {
+    // Denied MSRs not exported by Linux: IA32_FEATURE_CONTROL and IA32_MCG_CTL
+    if index == MSR_IA32_FEATURE_CONTROL || index == MSR_IA32_MCG_CTL {
+        return false;
+    };
+    ALLOWED_MSR_RANGES.iter().any(|range| range.contains(index))
+}
+
+/// Returns the list of supported, serializable MSRs.
+///
+/// # Arguments
+///
+/// * `kvm_fd` - Structure that holds the KVM's fd.
+pub fn supported_guest_msrs(kvm_fd: &Kvm) -> Result<Msrs> {
+    let mut msr_list = kvm_fd
+        .get_msr_index_list()
+        .map_err(Error::GetSupportedMSR)?;
+
+    msr_list.retain(|msr_index| msr_should_serialize(*msr_index));
+
+    let mut msrs =
+        Msrs::new(msr_list.as_fam_struct_ref().nmsrs as usize).map_err(|_| Error::CreateMsrs)?;
+    let indices = msr_list.as_slice();
+    let msr_entries = msrs.as_mut_slice();
+    // We created the msrs from the msr_list. If the size is not the same,
+    // there is a fatal programming error.
+    assert_eq!(indices.len(), msr_entries.len());
+    for (pos, index) in indices.iter().enumerate() {
+        msr_entries[pos].index = *index;
+    }
```

Comment:
Too bad that the logic of constructing `Msrs` cannot go into the constructor itself due to bindgen, and instead leaks into the client code :(